src/snowflake/snowpark/relational_grouped_dataframe.py

```diff
@@ -196,35 +196,59 @@ def _to_df(self, agg_exprs: List[Expression]) -> DataFrame:
+    ) -> DataFrame:
+        """Returns a :class:`DataFrame` with computed aggregates. See examples in :meth:`DataFrame.group_by`.
 
-        Valid input:
+        Args:
+            exprs: A variable length arguments list where every element is
 
-            - A Column object
-            - A tuple where the first element is a column and the second element is a name (str) of the aggregate function
-            - A list of the above
-        """
-        if not isinstance(exprs, (list, tuple)):
-            exprs = [exprs]
+                - A Column object
+                - A tuple where the first element is a column object or a column name and the second element is the name of the aggregate function
+                - A list of the above
+
+                or a ``dict`` maps column names to aggregate function names.
 
+        Note:
+            The name of the aggregate function to compute must be a valid Snowflake `aggregate function
+            <https://docs.snowflake.com/en/sql-reference/functions-aggregation.html>`_.
+
+        See also:
+            - :meth:`DataFrame.agg`
+            - :meth:`DataFrame.group_by`
+        """
+        exprs = parse_positional_args_to_list(*exprs)
         agg_exprs = []
-        for e in exprs:
-            if isinstance(e, Column):
-                agg_exprs.append(e._expression)
-            elif (
-                isinstance(e, tuple)
-                and isinstance(e[0], Column)
-                and isinstance(e[1], str)
-            ):
-                agg_exprs.append(_str_to_expr(e[1])(e[0]._expression))
-            else:
-                raise TypeError("Invalid input types for agg()")
+        if len(exprs) > 0 and isinstance(exprs[0], dict):
+            for k, v in exprs[0].items():
+                if not (isinstance(k, str) and isinstance(v, str)):
+                    raise TypeError(
+                        "Dictionary passed to DataFrame.agg() or RelationalGroupedDataFrame.agg() "
+                        f"should contain only strings: got key-value pair with types {type(k), type(v)}"
+                    )
+                agg_exprs.append(_str_to_expr(v)(Column(k)._expression))
+        else:
+            for e in exprs:
+                if isinstance(e, Column):
+                    agg_exprs.append(e._expression)
+                elif (
+                    isinstance(e, (list, tuple))
+                    and len(e) == 2
+                    and isinstance(e[0], (Column, str))
+                    and isinstance(e[1], str)
+                ):
+                    col_expr = (
+                        e[0]._expression
+                        if isinstance(e[0], Column)
+                        else Column(e[0])._expression
+                    )
+                    agg_exprs.append(_str_to_expr(e[1])(col_expr))
+                else:
+                    raise TypeError(
+                        "List passed to DataFrame.agg() or RelationalGroupedDataFrame.agg() should "
```

Comment:
Previous error messages are kinda inconsistent, and hard to be reused after this refactoring, so I changed it a little bit. But I will not regard it as a breaking change. But please comment if you have different opinions .

Reply:
I don't think it's a break change either.